### PR TITLE
S23 CSS Theme Palette Implementation

### PR DESCRIPTION
**Pull request by Centerville1**
First progress on converting colors to use CSS variables and the theme palette.

### CODE REVIEW

**Comment by Centerville1** [comment]
@EjPlatzer after talking with Dr. Tuck, he suggested we try and merge our progress into develop as in smaller parts as we go, to try to reduce merge conflicts with everyone else's changes (if we waited until we were finished).  So let me know if you see any changes needed, we've only changed a few components over so far, and some of those probably need to be converted to CSS later, instead of using in-line.

File: src/_vars.scss
Comment (EjPlatzer) [inline src/_vars.scss:98]: This seems unnecessary, since you still have the original values above.

File: src/app.global.scss
Comment (EjPlatzer) [inline src/app.global.scss:64]: Is this supposed to be `main` instead of `train`?
```suggestion
    color: var(--mui-palette-secondary-main) !important;
```

File: src/components/GordonDialogBox/GordonDialogBox.module.scss
Comment (EjPlatzer) [inline src/components/GordonDialogBox/GordonDialogBox.module.scss:3]: You created this file, but didn't edit the GordonDialogBox component at all? Or did you not commit the changes to the component yet?

File: src/components/GordonDialogBox/GordonDialogBox.module.scss
Comment (EjPlatzer) [inline src/components/GordonDialogBox/GordonDialogBox.module.scss:18]: Also, since this is a CSS module, we don't the very specific and long naming scheme.

It would be easier and clearer to flatten the CSS selectors and give them simpler, more clear names, e.g.
```suggestion
.dialog {
  display: grid;
  grid-template-rows: auto auto auto;
  overflow-y: auto;
}

.title {
    background-color: var(--mui-palette-primary-main);
    color: $neutral-white;
    text-align: center;
}

.content :global(.MuiDialogTitle-root) {
    padding: 8px 20px 8px 16px;
}

.actions {
    justify-content: space-evenly;
}
```

File: src/components/Header/Header.module.scss
Comment (EjPlatzer) [inline src/components/Header/Header.module.scss:68]: I think you missed the `$neutral-white` at the bottom of this component.

File: src/components/Header/index.jsx
Comment (EjPlatzer) [inline src/components/Header/index.jsx:143]: I believe this is unnecessary, since `primary` is the default value of the `color` prop for `AppBar`.

File: src/components/NavButton/index.jsx
Comment (EjPlatzer) [inline src/components/NavButton/index.jsx:33]: What was the reasoning behind applying the link styles to these? I'm not against the change necessarily, but I think it's outside the scope of this PR - this is a change to the expected styles of an element, whereas this PR should mostly be pure refactoring with on actual changes to the page.

File: src/components/Profile/components/Identification/Identification.module.scss
Comment (EjPlatzer) [inline src/components/Profile/components/Identification/Identification.module.scss:30]: I know this file is massive, but it seems to me that you ought to do an entire component in one go, rather than replacing it piece meal.

File: src/components/Profile/components/Identification/index.jsx
Comment (EjPlatzer) [inline src/components/Profile/components/Identification/index.jsx:67]: I think you missed a reference to `gordonColors` on line 76 below.

File: src/components/Profile/components/Identification/index.jsx
Comment (EjPlatzer) [inline src/components/Profile/components/Identification/index.jsx:54]: For this component to actually be ready for the dark mode switch, all of these hard-coded `white` colors will also need to be replaced. I'm not saying you need to do that now, but it seems pointless to change only part of a component's styles, and then have to come back to it again later.

File: src/components/Profile/components/MembershipsList/components/MembershipInfoCard/index.jsx
Comment (EjPlatzer) [inline src/components/Profile/components/MembershipsList/components/MembershipInfoCard/index.jsx:21]: We definitely shouldn't add **new** inline styles - and also this one is hard-coding the wrong value. I know you probably didn't add this, but it needs to be removed before we merge.

File: src/views/About/index.jsx
Comment (EjPlatzer) [inline src/views/About/index.jsx:18]: This is now unused in the file, please delete it.

File: src/views/Home/components/CLWCreditsDaysLeft/index.jsx
Comment (EjPlatzer) [inline src/views/Home/components/CLWCreditsDaysLeft/index.jsx:28]: We probably don't want to commit this file as is, since this is not an acceptable final solution. Please undo these changes for now until we find a working solution for chart.js

File: src/views/Timesheets/components/ShiftItem/index.jsx
Comment (EjPlatzer) [inline src/views/Timesheets/components/ShiftItem/index.jsx:29]: This is hard-coding the light theme, we shouldn't merge it.

File: src/views/Timesheets/index.jsx
Comment (EjPlatzer) [inline src/views/Timesheets/index.jsx:42]: As above, we shouldn't merge anything that hard-codes the light theme.

File: src/views/PeopleSearch/index.tsx
Comment (EjPlatzer) [inline src/views/PeopleSearch/index.tsx:13]: This is unused in this file, please remove it.

File: src/views/Home/components/NewsCard/index.jsx
Comment (EjPlatzer) [inline src/views/Home/components/NewsCard/index.jsx:16]: Is this necessary? The Card should have default styles, so this is either irrelevant or else you're changing the styles of this card specifically, which I don't think is in the scope of this PR.

File: src/views/Home/components/DiningBalance/index.jsx
Comment (EjPlatzer) [inline src/views/Home/components/DiningBalance/index.jsx:19]: As above, this is not a good final solution, so I don't think we should merge it.

**Review by EjPlatzer — CHANGES_REQUESTED** [review]
Great work on this y'all! Really exciting to see so much progress toward a unified site with editable theming!

I left several comments mostly on small cleanups to get this high quality before merging.

File: src/components/GordonDialogBox/GordonDialogBox.module.scss
Comment (Centerville1) [reply src/components/GordonDialogBox/GordonDialogBox.module.scss:3]: This might have been created as part of Sharon's branch, I haven't looked at this in detail yet, so correct, this is unfinished

File: src/components/Header/Header.module.scss
Comment (Centerville1) [reply src/components/Header/Header.module.scss:68]: Yeah, I haven't gone through header fully yet(just with my initial experimentation, but this will still work for now, will be changed when we get to that component.

File: src/components/Header/index.jsx
Comment (Centerville1) [reply src/components/Header/index.jsx:143]: Correct, yep, I'll fix that, again, this is still a mix of legacy and new code at this point, we have a document tracking which changes we have finished, and are doing internal code reviews to try to ensure completeness as much as possible.

File: src/components/Profile/components/Identification/Identification.module.scss
Comment (Centerville1) [reply src/components/Profile/components/Identification/Identification.module.scss:30]: Yep, most of the individual changes were just part of my experimentation, we're now going through whole components at once to change, and since this PR I've actually fully completed this component, so I'll include it in the next merge.

File: src/components/Profile/components/Identification/index.jsx
Comment (Centerville1) [reply src/components/Profile/components/Identification/index.jsx:54]: Yep, those changes will be in the next PR.  Sorry this first one is kind of piecemeal, most of the changes here are just from my experimentation to learn how to do this.  Now that all three of us are intentionally working on whole components, future PRs will be much more organized.

Comment (EjPlatzer) [reply src/components/Profile/components/Identification/index.jsx:54]: No worries, that makes sense.

File: src/components/Profile/components/Identification/index.jsx
Comment (Centerville1) [reply src/components/Profile/components/Identification/index.jsx:67]: Yep, that's already been changed since this PR, it'll be in the next one

File: src/components/Profile/components/MembershipsList/components/MembershipInfoCard/index.jsx
Comment (Centerville1) [reply src/components/Profile/components/MembershipsList/components/MembershipInfoCard/index.jsx:21]: Fair enough, yeah, I think this was from before I was working on it, so I'll revert it until we get to this file for real

File: src/views/About/index.jsx
Comment (Centerville1) [reply src/views/About/index.jsx:18]: getColor was still being used, but it was unneeded, so I'll change it.

File: src/views/Home/components/CLWCreditsDaysLeft/index.jsx
Comment (Centerville1) [reply src/views/Home/components/CLWCreditsDaysLeft/index.jsx:28]: Fair enough, I'll undo those for now :+1:

File: src/views/Home/components/NewsCard/index.jsx
Comment (Centerville1) [reply src/views/Home/components/NewsCard/index.jsx:16]: Yep, it is necessary, the card by default gets a color from the MUI palette, which isn't changed by out theme palette.  For all of the cards, I'm having to add a className to them to reference the CSS theme variables for their background colors...

File: src/views/Timesheets/components/ShiftItem/index.jsx
Comment (Centerville1) [reply src/views/Timesheets/components/ShiftItem/index.jsx:29]: Yeah, I think this is legacy from Sharon's work, reverting for now :+1:

File: src/views/Timesheets/index.jsx
Comment (Centerville1) [reply src/views/Timesheets/index.jsx:42]: Reverting for now :+1:

File: src/_vars.scss
Comment (Centerville1) [reply src/_vars.scss:98]: Yeah, sorry, meant to get rid of that

File: src/components/GordonDialogBox/GordonDialogBox.module.scss
Comment (Centerville1) [reply src/components/GordonDialogBox/GordonDialogBox.module.scss:3]: We'll look at this in detail when we get to it... is it okay to leave in for now?

File: src/components/GordonDialogBox/GordonDialogBox.module.scss
Comment (Centerville1) [reply src/components/GordonDialogBox/GordonDialogBox.module.scss:18]: Yeah, I know you mentioned that would be good to change, now that we use modules.  I'll try to work on that at some point, but that'll be a bit of a project.

File: src/components/NavButton/index.jsx
Comment (Centerville1) [reply src/components/NavButton/index.jsx:33]: Well, we tested with dark mode setting silly colors just to see what isn't referencing it right, and the cards are set to a default dark gray, rather than the yellow we had set in our dark mode for testing... so, it didn't work...

Comment (Centerville1) [reply src/components/NavButton/index.jsx:33]: Because of that, since this PR I've made changes to a bunch of cards like that, which will be included in future commits for the above reasons, unless there's something I'm missing

File: src/views/Home/components/DiningBalance/index.jsx
Comment (Centerville1) [reply src/views/Home/components/DiningBalance/index.jsx]: Fair , I removed all uses of getComputedStyle for now

File: src/components/GordonConfetti/index.jsx
Comment (Centerville1) [reply src/components/GordonConfetti/index.jsx]: Good point, I've reverted this for now until we get a chance to look at it in detail.

Comment (Centerville1) [reply src/components/GordonConfetti/index.jsx]: Also, as I said in other places, our future commits will be much more organized, this just has a lot of initial work and experiments, which will be standardized in the future.

File: src/app.jsx
Comment (Centerville1) [reply src/app.jsx]: Done 👍

File: src/components/GordonDialogBox/GordonDialogBox.module.scss
Comment (russtuck) [reply src/components/GordonDialogBox/GordonDialogBox.module.scss:3]: If it's not used, I don't think we should add it.  It will require attention to delete later, and distract people looking for things from what they really want.  Save it somewhere (or just leave it in whatever branch it came from).  But don't merge it into develop.

Comment (EjPlatzer) [reply src/components/GordonDialogBox/GordonDialogBox.module.scss:3]: I think that it would be better to remove it from the pull request so that we aren't merging unused code that will need to be reviewed later.

File: src/_vars.scss
Comment (russtuck) [inline src/_vars.scss:10]: As we discussed in person, every color needs a (single) name.  That name should be used in the theme definitions.  If you want to delete this name eventually, then create a new better name and use that.  (In fact, use the good name to define the old name, if possible).   In general, each hex code should only appear once anywhere.

File: src/_vars.scss
Comment (russtuck) [inline src/_vars.scss:11]: Please find all uses of this hex code, at least in this pull request, and replace them with the appropriate name.  Do the same for all the rest of the hex codes below.  (Exception: if there are two logically distinct names and uses for the same hex code, presumably used in different contexts, that's probably OK.)

File: src/views/About/index.jsx
Comment (EjPlatzer) [inline src/views/About/index.jsx:103]: Two things here:
1. You appear to have switched the color of this `Button` from cyan to blue.
2. You shouldn't need to pass this as an inline style, you can just pass the `color` prop to the button.

File: src/_vars.scss
Comment (russtuck) [inline src/_vars.scss:20]: Please put each comment above its color, unless it is very short, so there are no lines over 80 characters long.

File: src/views/Home/components/NewsCard/index.jsx
Comment (EjPlatzer) [reply src/views/Home/components/NewsCard/index.jsx:16]: Hmm, that seems to me like you might've configured the dark theme wrong then. As I understand it, MUI has a default dark theme which should effect all components when you switch themes, so you don't need to, for example, restyle all MUI components (which kinda defeats the purpose of using MUI in the first place).

Adding a class to EVERY `Card` component in the codebase is not an acceptable solution, so this might need further investigation.

File: src/_vars.scss
Comment (russtuck) [inline src/_vars.scss:48]: Good comment, but please say it in a way that will make sense a year from now (after the theme changes are done and forgotten)

File: src/_vars.scss
Comment (russtuck) [reply src/_vars.scss]: Why is it still here?

**Review by EjPlatzer — CHANGES_REQUESTED** [review]

File: src/app.global.scss
Comment (russtuck) [reply src/app.global.scss]: Please commit the suggestion.

**Review by russtuck — COMMENTED** [review]
Left a few comments, and as discussed, will hold off reviewing more since you're working on it and Evan is reviewing.

File: src/_vars.scss
Comment (Centerville1) [reply src/_vars.scss:10]: As we discussed, the _vars file will be deprecated by this update, and is only still here to ensure backwards compatibility while we transition.  Because of that, there's no need that I can see to updated the colors here with a different naming scheme.

File: src/_vars.scss
Comment (Centerville1) [reply src/_vars.scss:11]: See below comment for why outdated

File: src/_vars.scss
Comment (Centerville1) [reply src/_vars.scss:20]: We don't actually need these comments any more, so I'll just remove them...

File: src/_vars.scss
Comment (Centerville1) [reply src/_vars.scss:48]: I think the comments beneath this line are fairly clear, except for a typo I left, I'll commit the fixed typo, hopefully it's clear enough

File: src/components/GordonDialogBox/GordonDialogBox.module.scss
Comment (Centerville1) [reply src/components/GordonDialogBox/GordonDialogBox.module.scss:3]: So, we'll probably be adding a CSS file to Gordon Dialog Box anyway...  so the attention to check it later will already be given.  Would it be sufficient to add a comment to the CSS file for now, since it will likely be used later @EjPlatzer @russtuck ?

File: src/views/Home/components/NewsCard/index.jsx
Comment (Centerville1) [reply src/views/Home/components/NewsCard/index.jsx:16]: So, MUI does have a default dark theme, which works, but, at least the way I set up the themes, our theme doesn't override the card color from MUI.  I don't know if you have any idea what I might need to change to make our dark mode override the MUI properly.

One thing I do know, a lot of MUI components like buttons have color props that are set to MUI them setting, like color='primary,' and they DO change with our defined theme colors...  For testing purposes, I set all of the dark theme colors to yellow, so we can easily see which components have been changed to reference the correct theme CSS variables...  Buttons which have MUI color props do turn yellow in dark mode, but the cards stay the default MUI dark mode dark gray color.

I don't know if that helps you understand this at all @EjPlatzer , I'll try to read up on how all these themes interact, because I clearly don't understand it fully...

Comment (Centerville1) [reply src/views/Home/components/NewsCard/index.jsx:16]: In the meantime, I have been changing other cards in this way, which seems to work fine, but maybe I don't understand the problem with this.  
Is the problem that defining a CSS class for them overrides all the default MUI styles, thus defeating the purpose?  
I guess I would expect that defining a CSS class only overrides the MUI styles for CSS properties that we include manually, but that might be a faulty assumption on my part...

Comment (EjPlatzer) [reply src/views/Home/components/NewsCard/index.jsx:16]: Maybe a clearer way to say this is that, rather than overriding the MUI default again each time we use the MUI component (`Card` in this case), we should override the component's styling once at the theme level so that every `Card` across the entire site automatically gets overridden in a consistent way without us having to do any extra work in each place that `Card` is used.

I think perhaps the piece that you're missing is that all of the MUI theme is configurable, not just the palette. So if you want `Card` to have different styles, you can declare the `Card` styles for each theme mode inside the base theme, and it will apply to every `Card`.

The question I would ask is - do you need to change the `Card` styling? You mentioned you're using some yellow colors just for testing, but you only applied those colors to the palette colors, right? So it's expected that the MUI components that don't reference our palette colors are still using the default dark mode theme. Is there something wrong with the default dark mode theme? If so, you should change it at the theme level as outlined above. If not, you shouldn't change it all - just let the theme be as is.

Does that make sense?

Comment (Centerville1) [reply src/views/Home/components/NewsCard/index.jsx:16]: Ah, got it, that is helpful.  I think we probably want the cards to reference our palette colors, because we'll probably want to customize the color it chooses, especially in dark mode, rather than the default... but yes, that makes total sense that there is an easier way that I didn't understand, thanks!

File: src/components/ThemeModeSwitcher/index.jsx
Comment (amos-cha) [inline src/components/ThemeModeSwitcher/index.jsx:18]:
```suggestion
  //Need to figure out how to properly reference the theme provided by the CssVarsProvider context.
  console.log('Called mode switcher');
```

ditto

File: src/components/ThemeModeSwitcher/index.jsx
Comment (amos-cha) [inline src/components/ThemeModeSwitcher/index.jsx:25]:
```suggestion
  useEffect(() => {
    setMounted(true);
  }, []);
  if (!mounted) {
    return null;
  }
  ```
  In what case will this component not be mounted?

**Review by amos-cha — CHANGES_REQUESTED** [review]

File: src/components/ThemeModeSwitcher/index.jsx
Comment (Centerville1) [reply src/components/ThemeModeSwitcher/index.jsx:25]: Probably never, that mode switch button was just copied from the MUI docs, so it's probably designed to cover more cases than we have...  I'll remove the useEffect

File: src/components/GordonDialogBox/GordonDialogBox.module.scss
Comment (Centerville1) [reply src/components/GordonDialogBox/GordonDialogBox.module.scss:3]: > So, we'll probably be adding a CSS file to Gordon Dialog Box anyway... so the attention to check it later will already be given. Would it be sufficient to add a comment to the CSS file for now, since it will likely be used later @EjPlatzer @russtuck ?

@EjPlatzer thoughts on this?

Comment (EjPlatzer) [reply src/components/GordonDialogBox/GordonDialogBox.module.scss:3]: > > So, we'll probably be adding a CSS file to Gordon Dialog Box anyway... so the attention to check it later will already be given. Would it be sufficient to add a comment to the CSS file for now, since it will likely be used later @EjPlatzer @russtuck ?
> 
> @EjPlatzer thoughts on this?

If you're going to add it later, than it can be included in that pull request. It creates a needlessly confusing history for the file to have it created in git before it's referenced anywhere in the project. I think it should be removed before this PR is merged. We should, in general, strive for clean pull requests that are self-contained and leave the code cleaner than it was before.

Comment (Centerville1) [reply src/components/GordonDialogBox/GordonDialogBox.module.scss]: Okay, I've deleted it...

File: src/components/NavButton/index.jsx
Comment (EjPlatzer) [reply src/components/NavButton/index.jsx:33]: Ok, I see what you mean now. In one sense, the current code is correct because these aren't all links. But in a more pragmatic sense it probably does make sense to style them all consistently.

File: src/components/NavButton/index.jsx
Comment (EjPlatzer) [inline src/components/NavButton/index.jsx:44]: This added class is redundant with the class on the `NavLink`. You should only need one of these.

File: src/views/About/index.jsx
Comment (Centerville1) [reply src/views/About/index.jsx:103]: Good point, I've made that change :+1:

File: src/components/ThemeModeSwitcher/index.jsx
Comment (EjPlatzer) [inline src/components/ThemeModeSwitcher/index.jsx:24]: This is actually technically using stale data to update `setMode`. The right way to do this kind of conditional update based on previous state in React is with the callback version of `useState`:
```suggestion
      onClick={() => setMode(prevMode => prevMode === 'light' ? 'dark' : 'light')}
```

File: src/components/NavButton/index.jsx
Comment (Centerville1) [reply src/components/NavButton/index.jsx:44]: So, the link buttons were still displaying as different colors again when I only had them on the NavLink component, I can remove them from the navLink component, which seems to work from testing, so I'll commit that for review.

File: src/components/ThemeModeSwitcher/index.jsx
Comment (Centerville1) [reply src/components/ThemeModeSwitcher/index.jsx:24]: Interesting, this code was from the MUI docs for this feature, so I guess they've got some room for improvement, I'll make that change :slightly_smiling_face:

Comment (EjPlatzer) [reply src/components/ThemeModeSwitcher/index.jsx:24]: @Centerville1 It's possible that the custom hook doesn't have a callback format.

Your code is also not correct, so that would be the issue if the callback does work. The inner arrow function for the callback doesn't actually return anything. The way that arrow functions work, if you include braces after the arrow, then the function body is a statement that returns nothing unless you use an explicit return statement. Conversely, if you don't use braces for the body of the function, it's an expression, meaning that whatever value the expression evaluates to  is the value returned by the function.

For example, this logs `undefined` because the body of `addOne` is a statement with no `return` statement.
```javascript
const addOne = (x) => { x + 1 };
console.log(addOne(1));
```
Whereas this logs `2`
```javascript
const addOne = (x) => x + 1;
console.log(addOne(1));
```

Comment (Centerville1) [reply src/components/ThemeModeSwitcher/index.jsx:24]: > @Centerville1 It's possible that the custom hook doesn't have a callback format.
> 
> Your code is also not correct, so that would be the issue if the callback does work. The inner arrow function for the callback doesn't actually return anything. The way that arrow functions work, if you include braces after the arrow, then the function body is a statement that returns nothing unless you use an explicit return statement. Conversely, if you don't use braces for the body of the function, it's an expression, meaning that whatever value the expression evaluates to is the value returned by the function.
> 
> For example, this logs `undefined` because the body of `addOne` is a statement with no `return` statement.
> 
> ```js
> const addOne = (x) => { x + 1 };
> console.log(addOne(1));
> ```
> 
> Whereas this logs `2`
> 
> ```js
> const addOne = (x) => x + 1;
> console.log(addOne(1));
> ```

I tried it with and without brackets, I only just added the brackets when it didn't work without them.  I couldn't find anything online about useColorScheme having a callback function, although tbf there isn't that much documentation in general.

Comment (EjPlatzer) [reply src/components/ThemeModeSwitcher/index.jsx:24]: Gotcha, then it probably doesn't have a callback form, which is fine. The way you had it before isn't awful.

Comment (Centerville1) [reply src/components/ThemeModeSwitcher/index.jsx]: Thanks to Amos we were able to do it a little simpler, using ternary operators to do it, so should be a little simpler.

File: src/components/ThemeModeSwitcher/index.jsx
Comment (EjPlatzer) [inline src/components/ThemeModeSwitcher/index.jsx:6]: This looks like an unused import. Was it added purposefully?

Comment (Centerville1) [reply src/components/ThemeModeSwitcher/index.jsx:6]: Nope, got automatically added when we were experimenting, I'll get rid of it :+1:

**Review by EjPlatzer — APPROVED** [review]
There are still some opportunities for cleanup, but given this is being actively worked on, I think we're at a place where this is ready to merge. Great work y'all!

**Comment by Centerville1** [comment]
> There are still some opportunities for cleanup, but given this is being actively worked on, I think we're at a place where this is ready to merge. Great work y'all!

Great, thanks Evan,

**Comment by Centerville1** [comment]
> There are still some opportunities for cleanup, but given this is being actively worked on, I think we're at a place where this is ready to merge. Great work y'all!

Great, thanks Evan!  Are you able to merge this, I don't have merge permissions, and you may need to force a merge anyway, since there are requested changes from Amos (which I did address)

**Pull request merged**
